User Profiles Table in Analytics MySQL DB

bradenmacdonald opened the pull request:
In order to support user-specific reports in the Insights dashboard, this PR takes the first step by making user data available within the MySQL DB.

Two [related](https://github.com/edx/edx-analytics-data-api/pull/80) [PRs](https://github.com/edx/edx-analytics-data-api-client/pull/21) update the data API to allow analytics software to access the data, and [a final PR for the Dashboard](https://github.com/edx/edx-analytics-dashboard/pull/338) displays it to the user.

**Changes**:
* Add user's "name" to the existing task that imports data from the LMS `auth_userprofile` table to Hive.
* Adds two new pipeline tasks:
  * `InsertToMysqlUserProfilesTask` which creates the new `user_profile` table in the analytics MySQL DB. This new table contains basic user information (ID, name, email, username, gender, age, education level, login date) extracted and joined from the existing `auth_user` and `auth_userprofile` Hive tables.
  * `EnrollmentSnapshotTask` which creates/updates a new MySQL table (`course_enrollment_snapshot`) to contain the most current enrollment information available.

These changes together allow us to create new Data API methods that can (a) get the profiles of all known users or (b) get the profiles of all users in a specific course.

**Partner Information**: Third-party open edX instance, not an edX partner.

**Test Instructions - OpenCraft devstack**:
One way to test is with the analytics devstack:

1. Setup an [analytics devstack](https://github.com/open-craft/edx-analytics-devstack). 
2. Modify your local LMS devstack to work with the new analytics devstack (See analytics devstack README).
3. Login to the analytics devstack as the `analytics` user.
4. Run the command:
   ```
   launch-task InsertToMysqlUserProfilesTask --local-scheduler
   ```
5. Wait for it to complete. To check if it worked, run `mysql -u root analytics --execute="SELECT * FROM user_profile;"`. You should get a result set showing all the users on your LMS devstack.

**Test Instructions - Combined devstack**:
Another way is the nascent combined devstack:

1. Setup a [combined devstack](https://gist.github.com/bradenmacdonald/8b3b9736bf2b5ec44bc0)
2. Clear out any existing pipeline repo: `vagrant ssh` then `sudo rm -r /var/lib/analytics-tasks/devstack/repo` (This is necessary since the ansible scripts won't force the use of a different upstream repo during the next step)
3. On your host, in the edx-analytics-pipeline folder (with the venv loaded), run these commands:
    
    ```
    export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
    remote-task --vagrant-path ../ --remote-name devstack --override-config ${PWD}/config/devstack.cfg --wheel-url $WHEEL_URL --repo https://github.com/open-craft/edx-analytics-pipeline.git --branch user-table-mysql --wait InsertToMysqlUserProfilesTask --local-scheduler
    ```
4. Wait for it to complete, then inside the devstack, run `mysql -u root reports --execute="SELECT * FROM user_profile;"` You should get a result set showing all the users on your LMS devstack.
5. Now, try the other new task, `EnrollmentSnapshotTask`, with this command:
    
    ```
    remote-task --vagrant-path ../ --remote-name devstack --override-config ${PWD}/config/devstack.cfg --wheel-url $WHEEL_URL --repo https://github.com/open-craft/edx-analytics-pipeline.git --branch user-table-mysql --wait ImportEnrollmentsIntoMysql --local-scheduler --interval-end $(date -v+1d +%Y-%m-%d) --n-reduce-tasks 1
    ```
    (If the `date` subcommand isn't working, try `date +%Y-%m-%d -d "tomorrow"` instead)
6. Then, to check that it worked, run `mysql -u root reports --execute="SELECT * FROM course_enrollment_snapshot;"` as any user within the devstack.

**Author Notes/questions**
* Does it make sense for this data to be imported this way and stored in the analytics MySQL DB?
* Is it necessary to add a migration or something to account for a new column being added to the auth_userprofile Hive table?
* How do I write tests for this?

mulby commented:
@bradenmacdonald same deal here - let me know when you are ready for engineering and/or product review

shnayder commented:
Hi @bradenmacdonald -- we clearly need this info in Insights, so +1 from a product perspective. I do think you asked a good question of copying the data from the LMS is the right approach.

As Gabe said, we can schedule a code review and possibly an arch discussion -- let us know when that makes sense on your end.

mulby commented:
A high level architecture thought - I think I like the idea of pulling the user data into the pipeline instead of using LMS APIs primarily because it allows us get more consistent views in insights. If we "snapshot" the data every night, it will be consistent across all of our metrics. If we are constantly calling the LMS API directly an astute instructor may notice that there are more (or less) users listed as enrolled in the course in the UI than there are counted by the enrollment activity chart (for example).

Unless there is a pressing need to have data that is more fresh in Insights - I would default to the position that we should pull it through the pipeline.

@jab5569 @brianhw - thoughts?

jab5569 commented:
Great point @Gabe. I tend to agree given that all our current and future
data sourcing/processing patterns are pulled through the pipeline and
stored in the result store. As well we're not seeing any forcing functions
for Insights which would cause us to reconsider this approach. In fact
we've already begun to code student_engagement though the pipeline.

So for consistency sake, I'd agree with Gabe's position.


On Wed, Aug 12, 2015 at 10:02 AM, Gabe Mulley <notifications@github.com>
wrote:

> A high level architecture thought - I think I like the idea of pulling the
> user data into the pipeline instead of using LMS APIs primarily because it
> allows us get more consistent views in insights. If we "snapshot" the data
> every night, it will be consistent across all of our metrics. If we are
> constantly calling the LMS API directly an astute instructor may notice
> that there are more (or less) users listed as enrolled in the course in the
> UI than there are counted by the enrollment activity chart (for example).
>
> Unless there is a pressing need to have data that is more fresh in
> Insights - I would default to the position that we should pull it through
> the pipeline.
>
> @jab5569 <https://github.com/jab5569> @brianhw
> <https://github.com/brianhw> - thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-analytics-pipeline/pull/141#issuecomment-130315405>
> .
>



-- 
John Baker
edX | Analytics Architect
141 Portland Street, 8th Floor
Cambridge, MA 02139

john@edx.org <mcoyne@edx.org>
http://www.edx.org

bradenmacdonald commented:
@mulby et al. Ok, this is ready for an engineering review :)

mtyaka commented:
@bradenmacdonald :+1: Code looks good, tests would be nice :)

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-759](https://openedx.atlassian.net/browse/OSPR-759) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've added yourself to the [AUTHORS](https://github.com/edx/edx-analytics-pipeline/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

sarina commented:
@bradenmacdonald can you add tests to this PR?

@mulby do tests run automatically on this repo?

mulby commented:
@sarina - Automatic tests are currently disabled on this repo. We have a ticket in our backlog to re-enable them but need to make some changes to ensure they can be run securely. We are planning on moving the unit testing to travis.

@bradenmacdonald we typically rely on acceptance tests for this type of change since it is very difficult to unit test a hive query.

bradenmacdonald commented:
@mulby @sarina I know, and I'd love to write some acceptance tests but so far I have no way to test/run them. Gabe, should I just try writing some tests blindly and let you guys run them for me, or is OpenCraft going to have to either set up our own EMR cluster or refactor the acceptance test runner as we'd discussed? I'm not sure how to move forward most efficiently.

mulby commented:
Yeah I was thinking about that as well. We will likely be tackling the story to refactor the tests for local execution in the next couple of weeks.

I expect this test is so similar to the others that it may be fine to just write it blindly and I can help tweak it (if necessary) by running it on my end. You also may be able to hack the acceptance tests so that they work for you without actually implementing the "correct" method for running locally.

bradenmacdonald commented:
Sounds good @mulby - thanks.

I have pushed some acceptance tests now, but I have not tested them.

brianhw commented on edx/analytics/tasks/user_profiles.py:
```diff
@@ -0,0 +1,85 @@
+        kwargs_for_db_import = {
+            'overwrite': False,
+            'import_date': self.import_date,
+            # Are other parameters needed?
```
nit: once you figure this out, remove the comment.

brianhw commented:
Code LGTM.  :+1:   Waiting on tests.  And please squash before merging.

bradenmacdonald commented on edx/analytics/tasks/database_imports.py:
```diff
@@ -261,6 +261,7 @@ def table_name(self):
     def columns(self):
         return [
             ('user_id', 'INT'),
+            ('name', 'STRING'),
```
@brianhw Thanks for reviewing this PR. I just want to double check on this line: Will this change to an existing Hive table cause any problems when it's merged? I'm not sure how schema changes in Hive work.

brianhw replied:
A fair question.  This will be a problem if we try to use older versions of the table data with these newer ones.  But I don't see it being as big a problem as making changes to Mysql tables, where we have to blow away the old tables and create new ones.  For these Hive tables, we are only loading the one partition, and as long as the data on disk from the Sqoop run matches the metadata definition used when loading the data into Hive, we should be fine.

bradenmacdonald commented:
@brianhw Do you know why the acceptance tests are still showing as "running" ?

mulby commented:
@bradenmacdonald I'm messing with the configuration a bit

bradenmacdonald commented:
Rebased, added Gabe's `db.py` fix commit 3f58cbfcb3979f44e46a12ff41d4821266ff8754, squashed, added myself to authors. Would you guys like to run the acceptance tests again, or is this ready to merge?

mulby commented:
@bradenmacdonald and @antoviaque - after extensive discovery on our own per-student analytics efforts, it's looking like we are going to use a variant of this patch. Specifically we will make use of the "user_profile" table, but will likely not include the enrollment snapshot table.

I'm closing this PR for now and will open a separate one that will focus exclusively on getting the user_profile table merged and will tag Braden as a reviewer.

antoviaque commented:
@mulby Thanks for the update! Also ping me at the same time on the new PR, if you haven't already, so I can take it into account for the next sprint scheduling (we have one sprint per week, starting every Monday FYI).
